hda: separation of l1 settings to power manager

fkwasowi opened the pull request:
Separating two new functions force and allow l1 to have the current state with separated functions in the power manager so that SOF can call these functions via IPC DMI_FORCE_L1_EXIT. Change related to the addition of a new parameter to force DMI L1 exit on IPC request.

Associated with: https://github.com/thesofproject/sof/pull/8561/files

teburd reviewed (changes requested):
This appears to be moving some ace specifics into power management for all of Zephyr while not actually changing the logic involved.

If soc specifics are needed in sof then those need to be exposed through the soc headers not through the general pm device header.

ceolin reviewed (changes requested):
This adsp specific logic and should never be part of Zephyr's device pm interface.

fkwasowi commented:
> This adsp specific logic and should never be part of Zephyr's device pm interface.

I changed the location of the function to soc/xtensa/intel_adsp/common/include/intel_adsp_ipc.h

fkwasowi committed:
hda: separation of l1 settings to new function

Separating two new functions force and allow l1
to have the current state with separated functions
in the ipc file so that SOF can call these
functions via IPC DMI_FORCE_L1_EXIT. Change related
to the addition of a new parameter to force
DMI L1 exit on IPC request.

Signed-off-by: Fabiola Kwasowiec <fabiola.kwasowiec@intel.com>

fkwasowi commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -206,6 +207,15 @@ void intel_adsp_ipc_send_message_emergency(const struct device *dev, uint32_t da
 	regs->idr = data | INTEL_ADSP_IPC_BUSY;
 }
 
+void intel_adsp_ipc_force_dmi_l0_state(void)
```
Corrected

teburd commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -465,7 +466,7 @@ void intel_adsp_hda_dma_isr(void)
 
 	if (clear_l1_exit) {
 #if CONFIG_SOC_SERIES_INTEL_ACE
```
Since the ifdefs make these no-ops in case of not ACE then the ifdef wrapping the call can be removed

fkwasowi replied:
Fixed

teburd reviewed (dismissed):
A remaining nit, but not a deal breaker, looks fine to me now

ceolin commented on drivers/dma/dma_intel_adsp_hda.c:
```diff
@@ -183,7 +184,7 @@ int intel_adsp_hda_dma_host_reload(const struct device *dev, uint32_t channel,
 
 #if CONFIG_DMA_INTEL_ADSP_HDA_TIMING_L1_EXIT
 #if CONFIG_SOC_SERIES_INTEL_ACE
```
You can remove this condition since you are already doing it in the implementation

fkwasowi replied:
Done

fkwasowi commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -206,6 +207,19 @@ void intel_adsp_ipc_send_message_emergency(const struct device *dev, uint32_t da
+#endif
+}
+
+void intel_adsp_ipc_allow_dmi_l1_state(void)
```
Corrected

ranj063 replied:
@fkwasowi what is DMA_l1? isnt dmi_l1 the right terminology to use?

fkwasowi replied:
Corrected

kv2019i commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -206,6 +207,19 @@ void intel_adsp_ipc_send_message_emergency(const struct device *dev, uint32_t da
+#ifdef CONFIG_SOC_SERIES_INTEL_ACE
+	ACE_DfPMCCH.svcfg &= ~(ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT);
+#endif
+}
```
The code looks otherwise good, but I'm not sure about the correct location. This is really a SoC-wide quirk that affects the HD DMA operation. So having it in soc/xtensa/intel_adsp/ makes sense, so this part is correct. But this doesn't really have much to do with ipc, so I'd rather put this into soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h 
... and keep it a no-op for non-ace platforms.

fkwasowi replied:
I moved the functions to the file you mentioned.

kv2019i reviewed (commented):
Code looks good but I think intel_adsp_hda is a better fit than ipc.h for this functionality.

kv2019i reviewed (dismissed):
Thanks @fkwasowi , looks good to me now.

marc-hb reviewed (dismissed):
> shouldn't these functions be called intel_adsp_ipc_allow_dma_l1_state and intel_adsp_ipc_allow_dma_l0_state ?

It's hard to believe anyone could make the same typo 6 times in a function name in a 10 lines long commit....

Is this maybe Direct Media Interface (PCI) link state?

In any case this is called "DMI L1" in https://github.com/thesofproject/sof/commit/173adc990c68 and everywhere else I looked so let's not introduce inconsistency. `west grep` saves lives.

cc: @RanderWang , @ranj063 

EDIT: there are also a few "HDMA L1" occurrences already but no "DMA L1" yet. Let's not yet another, 3rd variant.

fkwasowi commented:
> Direct Media Interface

Yes. It's about Direct Media Interface

kv2019i commented:
@marc-hb wrote:
>> shouldn't these functions be called intel_adsp_ipc_allow_dma_l1_state and intel_adsp_ipc_allow_dma_l0_state ?
> Is this maybe Direct Media Interface (PCI) link state?

I don't think it doesn't matter much here. The property we are setting is really tied to DMA (not typo) here. Behind the DMA, there will be DMI in the end, but from the point of view of the DSP, this difference has little practical meaning.

marc-hb commented:
When you run `west grep -i dmi.*l1`, you really don't want to miss one spot.

iganakov reviewed (approved):
Looks good to me

mwasko commented:
@nashif @ceolin can we proceed with merge?

marc-hb commented:
This fails to compile for LNL:

https://github.com/thesofproject/sof/actions/runs/7196263630/job/19600738125?pr=8618

```
-o zephyr/drivers/dma/CMakeFiles/drivers__dma.dir/dma_intel_adsp_hda.c.obj -c /zep_workspace/zephyr/drivers/dma/dma_intel_adsp_hda.c
In file included from /zep_workspace/zephyr/drivers/dma/dma_intel_adsp_hda.c:28:
/zep_workspace/zephyr/soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h: In function 'intel_adsp_force_dmi_l0_state':
/zep_workspace/zephyr/soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:448:30: error: 'ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT' undeclared (first use in this function)
  448 |         ACE_DfPMCCH.svcfg |= ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT;
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/zep_workspace/zephyr/soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:448:30: note: each undeclared identifier is reported only once for each function it appears in
/zep_workspace/zephyr/soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h: In function 'intel_adsp_allow_dmi_l1_state':
/zep_workspace/zephyr/soc/xtensa/intel_adsp/common/include/intel_adsp_hda.h:455:32: error: 'ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT' undeclared (first use in this function)
  455 |         ACE_DfPMCCH.svcfg &= ~(ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT);
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

marc-hb commented:
By using ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT in a common .h file, this commit 
introduces a new requirement:

=> now ADSP_FORCE_DECOUPLED_HDMA_L1_EXIT_BIT must **always** be defined, even when CONFIG_DMA_INTEL_ADSP_HDA_TIMING_L1_EXIT is false.

Is this new requirement OK?

Of course another question is: why is CONFIG_DMA_INTEL_ADSP_HDA_TIMING_L1_EXIT false on LNL? But that's a different question.

marc-hb commented:
> Of course another question is: why is CONFIG_DMA_INTEL_ADSP_HDA_TIMING_L1_EXIT false on LNL? But that's a different question.

5 months and probably a lot of wasted time later:

- #73733
